interfaces/builtin/opengl.go

```diff
@@ -183,6 +183,48 @@ unix (bind,listen) type=seqpacket addr="@cuda-uvmfd-[0-9a-f]*",
 # From https://bugs.launchpad.net/snapd/+bug/1862832
 /run/nvidia-xdriver-* rw,
 unix (send, receive) type=dgram peer=(addr="@var/run/nvidia-xdriver-*"),
+
+# Hybris support
+/{,var/}run/shm/hybris_shm_data rw, # FIXME: LP: #1226569 (make app-specific)
+# This is a LD_PRELOADed TLS padding library allowing use of both
+# Android 9 and glibc TLS slots at the same time. Only used on Halium 9.
+/usr/lib/@{multiarch}/libtls-padding.so  mr,
```

Comment:
This particular library resides on the host, which then gets picked up by snap-confine AFAICT.

In general thread-locals in a shared library are taken care of after initialization by `ld`, before the actual executable is run. Think calling constructors of a static variable, it's similar in style to how `ld` treats thread-locals.

Reply:
Just for the record, host files do not appear at that location. What you see here comes from the base snap.

Reply:
I am surprised then that snap-confine correctly stops complaining about not being able to map that library.

Reply:
If you're setting LD_PRELOAD in your environment, than the denial from s-c should be raised regardless of this interface, as s-c is running under a different profile. Now, setting LD_PRELOAD in the environment will not have an effect of the snap as IIRC LD_PRELOAD isn't carried over to the snap's environment. Now, if OTOH is being set within the snap, then yes, if you allow it explicitly in the file like here, the denial will go away even if the file doesn't exist. Can you paste the denial you saw?

Reply:
LD_PRELOAD is set in the host environment, yes. This results in two things:

- `ld` complaining during the preload itself: `ERROR: ld.so: object 'libtls-padding.so' from LD_PRELOAD cannot be preloaded (failed to map segment from shared object): ignored.`
- AppArmor giving the reason why:

```
[23609.467580] audit: type=1400 audit(1712049539.906:1289): apparmor="DENIED" operation="file_mmap" profile="/usr/lib/snapd/snap-confine" name="/usr/lib/aarch64-linux-gnu/libtls-padding.so" pid=33306 comm="snap-confine" requested_mask="m" denied_mask="m" fsuid=0 ouid=0
```

Of course the env var is set system-wide because we already need it on the host, but the confined snap also needs to inherit the padded TLS area.

Reply:
The denial for snap-confine is expected, since the PR does not modify the profile, but since it's functioning I'm not sure I'd update it. 
Within the snap, unless something sets LD_PRELOAD again, that environment variable will be empty, because ld.so will strip it out before switching to snap-confine. Are you setting it explicitly in the snaps you use for testing?

Reply:
I am not setting them within the Snap's environment and the the process' file mappings don't seem to reference the tls-padding library as well. I wonder if somehow the padded TLS area might be set up already in snap-confine and kept across exec()s, but I honestly doubt that.

Reply:
You can start a process in some test snap and then on the host either list its open files or inspect its memory map.

Reply:
That's what I did to no avail. Seems though my historical assumptions were wrong as the padding hack was a requirement for Halium/Android 10, not 9, and I don't have a H10 MediaTek device available at the moment.

On the other hand, could it be that GCC has also adapted to Android's TLS slot layout? In that case core20+ Snaps wouldn't necessarily need it. More infos here: https://reviews.llvm.org/D53906

Reply:
After giving it some time, I've come to the conclusion that we might still want to pass through the preload hack to the confined environment in case it's running on Touch specifically. Would that be feasible from your perspective?

Reply:
ld.so will strip out LD_PRELOAD before it switches execution to snap-confine, i.e. snap-confine will not even see that environment variable, so there's no chance of passing it to the snap. Perhaps you can load it from /etc/environment when executing in the snap, but it would not be automatic.

Reply:
I was more thinking of measuring whether it's running on Touch via other means (checking for existance of a Halium-only file) and recreating the environment variable. It might sound hacky at first but there really seems to be no easy other way.

Reply:
I presume you've ran some apps with this change built locally. Were they able to start and function properly?

Reply:
No I couldn't as my assumption about Halium 9 vs 10 was wrong. I thought one of my test devices was susceptible to the issue, whereas it was just Halium 10 MediaTek devices specifically (which seem to have TLS slots hardcoded in their bionic-based GL libs).